enhancements/ingress/aws-load-balancer-operator.md

```diff
@@ -0,0 +1,394 @@
+  DisabledAddons []AWSAddon // indicates which AWS addons should be disabled.
+}
+
+type AWSAddon string
+
+const (
+  AWSAddonShield AWSAddon = "AWSShield"
+  AWSAddonWAFv1 AWSAddon = "AWSWAFv1"
+  AWSAddonWAFv2 AWSAddon = "AWSWAFv2"
+)
+
+type DeploymentConfig struct {
+  // +optional
+  Replicas int
+}
+```
+
+
+### Implementation Details/Notes/Constraints
+
+The operator only reconciles a single instance of the custom resource because
+only [one
+instance](https://github.com/kubernetes-sigs/aws-load-balancer-controller/issues/2185)
+of the lb-controller can be run in a cluster. There can be multiple replicas
+with the same configuration with leader-election enabled. But multiple
+controllers cannot be started with different configurations.
+
+The lb-controller also requires that the subnets where the load balancers are
+provisioned have certain resource tags present on them. The operator can detect
+the subnets and tag them or the user can do this. This is specified with the
+_AutoTagSubnet_ field.
+
+The user can also specify the Ingress class which the lb-controller will
+reconcile. This could default to **_“alb”_**. This value is required because if
+an ingress class value is not passed to the controller it attempts to reconcile
+all Ingress resources which don’t have an ingress class annotation or where the
+ingress class field is empty. Passing the value of “alb” does not restrict the
+number of ingress classes which the controller can reconcile. This is described
+further in the [Parallel operation of Openshift
+Router](#parallel-operation-of-the-openshift-router-and-lb-controller) and
+lb-controller section.
+
+The _DeploymentConfig_ field contains any configuration required for the
+deployment. The number of replicas is the only field for now. But in future
+versions other fields from the
+[PodSpec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#podspec-v1-core)
+could also be exposed here.
+
+Finally, WAFs and DDos prevent configurations are attached to the load balancer
+through annotations specified on the Ingress resource. But users might want to
+limit which features are available and the `DisabledAddons` array can be used to
+specify which features/addons have to be disabled. By default all the addons
+will be enabled. They are disabled by setting the addon CLI flag to false. When
+an addon which was previously enabled is disabled the controller does not remove
+the existing addon attachment from the provisioned load balancers.
+
+The operator will also create a
+[CredentialRequest](https://docs.openshift.com/container-platform/4.9/rest_api/security_apis/credentialsrequest-cloudcredential-openshift-io-v1.html#credentialsrequest-cloudcredential-openshift-io-v1)
+on behalf of the lb-controller and mount the minted credentials in the
+_Deployment_ of the controller. This means that the operator will only work on
+OCP/OKD but this is sufficient for the initial release.
+
+
+### Risks and Mitigations
```

Comment:
Just to track what we discussed in Slack. Hope it can be resolved, in that case we can just resolve this comment.

I was wondering where we can get `VPCID` needed for the OCP deployment as it doesn't allow the access to AWS metadata server although [it will be needed](https://kubernetes-sigs.github.io/aws-load-balancer-controller/v2.3/deploy/installation/#using-metadata-server-version-2-imdsv2).
I could not find it in [Infrastructure resource](https://github.com/openshift/api/blob/8779abfbd5714ea73487699cbc6fc6e8e5f79158/config/v1/types_infrastructure.go#L340-L366). Other 2 things which will be needed by the operator are there though: [region](https://github.com/openshift/api/blob/8779abfbd5714ea73487699cbc6fc6e8e5f79158/config/v1/types_infrastructure.go#L351), [cluster name](https://github.com/openshift/api/blob/8779abfbd5714ea73487699cbc6fc6e8e5f79158/config/v1/types_infrastructure.go#L55).

Reply:
Added a section on the changes needed in the Installer.